File: Data Structures & Algorithms/trapping-rain-water/submission-7.swift
class Solution {
    func trap(_ height: [Int]) -> Int {

        let n = height.count

        var prefix:[Int] = Array(repeating: 0, count: n)
        var suffix: [Int] = Array(repeating: 0, count: n)


        prefix[0] = height[0]

        suffix[n-1] = height[n-1]

        for i in 1 ..< n {
            prefix[i] = max(height[i], prefix[i-1])
        }

        for i in stride(from:n-2, through: 0, by: -1) {
            suffix[i] = max(height[i], suffix[i+1])
        }

       
        var total = 0
        var trapped = 0

        for i in 0..<n {
            trapped = abs(min(prefix[i],suffix[i]) - height[i])

            if trapped > 0 {
                total += trapped
            }
        }

        return total

    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-7.java
class Solution {
    public int[] productExceptSelf(int[] nums) {

        // int product = 0;
        int n = nums.length;

        // int[] res = new int[n];

        // for(int i=0;i<n;i++){
        //     product = 1;

        //     for(int j=0;j<n;j++){

        //         if(i!=j){
        //             product*= nums[j];
        //         }
        //     }
        //     res[i] = product;


        // }

        // return res;

        int[] prefix = new int[n];
        prefix[0] = 1;

        int[] suffix = new int[n];
        suffix[n-1] = 1;
        int[] res = new int[n];

        for(int i=1;i<n;i++){
            prefix[i] = prefix[i-1]*nums[i-1];
        }

         for(int i=n-2;i>=0;i--){
            suffix[i] = suffix[i+1]*nums[i+1];
        }

         for(int i=0;i<n;i++){
            res[i] = prefix[i]*suffix[i];
         }

         return res;


        
    }
}  
